File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    count = 0
    arr = [[] for _ in range(n)]
    visited = [False] * n
    
    def bfs(i):
        queue = deque()
        queue.append(i)
        visited[i] = True
        
        while queue:
            cur = queue.popleft()
            
            if arr[cur]:
                for nxt in arr[cur]:
                    if not visited[nxt]:
                        queue.append(nxt)
                        visited[nxt] = True
                        
    for i in range(n):
        for j in range(n):
            if i != j and computers[i][j] == 1:
                arr[i].append(j)
            
    for i in range(n):
        if not visited[i]:
            bfs(i)
            count += 1
        
    return count

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] computers) {
        boolean[] visited = new boolean[n];
        List<List<Integer>> graph = new ArrayList<>();
        
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i != j && computers[i][j] == 1) {
                    graph.get(i).add(j);
                }
            }
        }

        int count = 0;

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                bfs(i, graph, visited);
                count++;
            }
        }

        return count;
    }

    private void bfs(int start, List<List<Integer>> graph, boolean[] visited) {
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(start);
        visited[start] = true;

        while (!queue.isEmpty()) {
            int cur = queue.poll();
            for (int next : graph.get(cur)) {
                if (!visited[next]) {
                    queue.offer(next);
                    visited[next] = true;
                }
            }
        }
    }
}
